File: 21-graphs3/6-bipartite-check/main.py
# Bipartite Graph Check using BFS - Python Example

from collections import deque

def is_bipartite(graph):
    color = {}
    for node in graph:
        if node not in color:
            queue = deque([node])
            color[node] = 0

            while queue:
                u = queue.popleft()
                for v in graph[u]:
                    if v not in color:
                        color[v] = 1 - color[u]
                        queue.append(v)
                    elif color[v] == color[u]:
                        return False
    return True

graph1 = {
    0: [1, 3],
    1: [0, 2],
    2: [1, 3],
    3: [0, 2]
}

graph2 = {
    0: [1, 2],
    1: [0, 2],
    2: [0, 1]
}

print("Graph1 is bipartite:", is_bipartite(graph1))  # True
print("Graph2 is bipartite:", is_bipartite(graph2))  # False

File: 21-graphs3/6-bipartite-check/main.cpp
// Bipartite Graph Check using BFS - C++ Example

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

bool isBipartite(const vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> color(n, -1);

    for (int i = 0; i < n; i++) {
        if (color[i] == -1) {
            queue<int> q;
            q.push(i);
            color[i] = 0;

            while (!q.empty()) {
                int u = q.front();
                q.pop();

                for (int v : graph[u]) {
                    if (color[v] == -1) {
                        color[v] = 1 - color[u];
                        q.push(v);
                    } else if (color[v] == color[u]) {
                        return false;
                    }
                }
            }
        }
    }
    return true;
}

int main() {
    vector<vector<int>> graph1 = {
        {1, 3},
        {0, 2},
        {1, 3},
        {0, 2}
    };

    vector<vector<int>> graph2 = {
        {1, 2},
        {0, 2},
        {0, 1}
    };

    cout << "Graph1 is bipartite: " << boolalpha << isBipartite(graph1) << endl;
    cout << "Graph2 is bipartite: " << boolalpha << isBipartite(graph2) << endl;

    return 0;
}
